taichi/transforms/offload.cpp

```diff
@@ -626,40 +546,32 @@ class FixCrossOffloadReferences : public BasicStmtVisitor {
       stmt_to_offloaded[copy.get()] = stmt_to_offloaded[stmt];
       stmt->set_operand(index, copy.get());
       stmt->insert_before_me(std::move(copy));
-      return true;
     }
 
-    if (op->is<PtrOffsetStmt>() &&
-        op->cast<PtrOffsetStmt>()->origin->is<AllocaStmt>()) {
-      auto alloca = op->as<PtrOffsetStmt>()->origin->as<AllocaStmt>();
-      if (local_to_global_offset.find(alloca) == local_to_global_offset.end())
-        return false;
-
-      auto global_temporary_stmt = Stmt::make<GlobalTemporaryStmt>(
-          local_to_global_offset[alloca], alloca->ret_type);
-      // TODO: offset may not be ConstStmt
-      auto copy_stmt = op->as<PtrOffsetStmt>()->offset->as<ConstStmt>()->copy();
-      auto ptr_offset_stmt = Stmt::make<PtrOffsetStmt>(
-          global_temporary_stmt.get(), copy_stmt.get());
-      stmt_to_offloaded[copy_stmt.get()] = stmt_to_offloaded[stmt];
-      stmt_to_offloaded[ptr_offset_stmt.get()] = stmt_to_offloaded[stmt];
-      stmt->set_operand(index, ptr_offset_stmt.get());
-      stmt->insert_before_me(std::move(copy_stmt));
-      stmt->insert_before_me(std::move(global_temporary_stmt));
-      stmt->insert_before_me(std::move(ptr_offset_stmt));
-      return true;
+    if (local_to_global_offset.find(op) == local_to_global_offset.end()) {
+      // For cases like ConstStmt
+      auto copy = op->clone();
+      stmt_to_offloaded[copy.get()] = stmt_to_offloaded[stmt];
+      stmt->set_operand(index, copy.get());
+      stmt->insert_before_me(std::move(copy));
```

Comment:
If this part can handle ConstStmt correctly, we should remove the code fragment `if (config.advanced_optimization) {  if (op->is<ConstStmt>()) ...` to avoid inserting the statement twice.
Also, it's better to give an example other than ConstStmt, otherwise `if (op->is<ConstStmt>())` is more intuitive to me.

Reply:
We had a discussion on this and feel like this could be a better/simpler approach: Regardless of `advanced_optimization`, we should *not* put Const into GlobalTmps. Instead, if OffloadedTask 2 references a Const defined in OffloadedTask 1, we can just clone that Const.

Reply:
Yeah but when `advanced_optimization=True`, the current code clone the const twice?

Reply:
https://github.com/taichi-dev/taichi/pull/2673/commits/87be94aee909d74b44319cbe1df7096e8286b05c